pkg/util/exec/exec_test.go

```diff
@@ -19,138 +19,148 @@ package exec
+		})
+	}
+}
+
+func TestRunShutdownError(t *testing.T) {
+	for _, s := range []os.Signal{syscall.SIGINT, syscall.SIGTERM} {
+		s := s // pin! see https://github.com/kyoh86/scopelint for why
+
+		t.Run(s.String(), func(t *testing.T) {
+			cmd := exec.Command("sleep", "10")
+
+			// Fake the actual signal handler. Signals sent to stopCh aren't actually sent to the process.
+			stopCh := make(chan os.Signal, 1)
+			revertSetupSignalHandler := overrideSetupSignalHandler(func(ctx context.Context) <-chan os.Signal {
+				return stopCh
+			})
+			defer revertSetupSignalHandler()
+
+			// We run the command in a goroutine as we need to signal the process, which results in an error.
+			errCh := make(chan error)
+			go func() {
+				defer close(errCh)
+				errCh <- Run(cmd, ioutil.StdStreams{In: new(bytes.Buffer), Out: new(bytes.Buffer), Err: new(bytes.Buffer)})
+			}()
+
+			stopCh <- s // Trigger exec.terminate()
+			close(stopCh)
+
+			// Verify the process shutdown and raised an error due to the signal we caught
+			for err := range errCh {
+				require.Equal(t, commonerrors.NewShutdownError(s), err)
 			}
-		}()
+		})
+	}
+}
 
-		Eventually(stdout.String).Should(Equal("running"))
+func TestRunSendsSIGTERMIfProcessStillRunningAfterStopSignal(t *testing.T) {
+	stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
+	stdio := ioutil.StdStreams{In: new(bytes.Buffer), Out: stdout, Err: stderr}
 
-		stopCh <- syscall.SIGINT
-		close(stopCh)
+	// Sleep for 2 seconds, which should be enough to know that the SIGTERM was issued
+	cmd := exec.Command("testdata/test_trap_sigterm.sh", "2")
 
-		var err error
-		Eventually(errCh, "5s", "100ms").Should(Receive(&err))
-		Expect(err).To(Equal(commonerrors.NewShutdownError(syscall.SIGINT)))
+	// Fake the actual signal handler. Signals sent to stopCh aren't actually sent to the process.
+	stopCh := make(chan os.Signal, 1)
+	revertSetupSignalHandler := overrideSetupSignalHandler(func(ctx context.Context) <-chan os.Signal {
+		return stopCh
 	})
-})
+	defer revertSetupSignalHandler()
+
+	// We run the command in a goroutine as we need to SIGTERM the process, which results in an error.
+	errCh := make(chan error)
+	go func() {
+		errCh <- Run(cmd, stdio)
+		close(errCh)
+	}()
+
+	// Wait for shell script to start
+	require.Eventually(t, func() bool {
+		return strings.Contains(stdout.String(), "running")
+	}, 1*time.Second, 10*time.Millisecond)
+
+	// Send a fake signal to our signal handler which invokes the stop channel
+	fakeSignal := syscall.Signal(0)
+	stopCh <- fakeSignal
+	close(stopCh)
+
+	// Wait for the err channel which means the process shutdown.
+	for err := range errCh {
+		require.Equal(t, commonerrors.NewShutdownError(fakeSignal), err)
+	}
+
+	// Verify the program received a SIGTERM because the process was still alive upon exec.terminate()
+	require.Equal(t, "SIGTERM caught!\n", stderr.String(), `invalid stderr running [%v]`, cmd)
```

Comment:
this is important to know because otherwise we don't know why we expect a kill to eventually occur.